core/src/test/java/io/grpc/util/GracefulSwitchLoadBalancerTest.java

```diff
@@ -79,6 +80,11 @@ public void setUp() {
     }
   }
 
+  @Nullable
+  public static String getCurrentPolicyForTest(GracefulSwitchLoadBalancer gracefulSwitchLb) {
```

Comment:
Why not just have a helper method in `XdsLoadBalancerTest.java` instead of putting it here?

Reply:
Because in xds package, it does not have access to the method `gracefulSwitchLb.getCurrentPolicyForTest()`. Otherwise need declare it as public.

Reply:
`XdsLoadBalancer` is more like a wrapper and the actual load balancing logic is delegated to the component `gracefulSwitchLb`, which is a `LoadBalancer`. So maybe `XdsLoadBalancer` should not leak detail of having a `GracefulSwitchLoadBalancer` but instead just a `LoadBalancer`. We could inject a fake `LoadBalancer` in test and are able to populate the current policy name from the fake `LoadBalancer` component.

Reply:
`XdsLoadBalancer` uses `GracefulSwitchLoadBalancer.swapTo()`, this is not something that can be faked or injected. It's only leaked to test. Otherwise it's hard to get current policy and current balancer because `GracefulSwitchLoadBalancer` hides it. Indirect verification is very hard because calling a method of `GracefulSwitchLoadBalancer` and then verify a mock delegate has called that method is wrong because that method may delegate to the pending policy, not the current policy.

Reply:
Discussed offline with @zhangkun83 who proposed: for testability, introducing new public APIs

```
// GracefulSwitchLoadBalancer will extend this class
public abstract class SwitchLoadBalancer extends LoadBalancer {
  public abstract void swapTo(oadBalancerProvider newLbProvider);
}

public abstract class SwitchLoadBalancerFactory {
  public abstract SwitchLoadBalancer newSwitchLoadBalancer(Helper helper);
}
```

and injecting fake `SwitchLoadBalancerFactory` for test.
@voidzcy @ejona86 your thought?

Reply:
I think I'd rather not. I'd prefer the tests to verify the actual LB behavior instead of the implementation.

Reply:
The current `XdsLoadBalancerTest` are not testing the actual LB behavior, but the child-policy switching behavior. I would prefer the verification of actual LB behavior in a separate integration-style test, which, however, won't cover all code paths. `XdsLoadBalancerTest` is a unit test for that class, and introducing those interfaces would make it easy.

As we expect `GracefulSwitchLoadBalancer` to be useful for other people, they might also appreciate mock-ability for their own tests.

Reply:
I see a lot of digging into `XdsLbState`. What would a `TestingSwitchLoadBalancer` do?

Reply:
AFAIK the digging only verifies child policy changes, which would be changed to `verify`s on `switchTo()`.

Reply:
Agree.

In current `XdsLoadBalancerTest`, we are verifying that the actual load balancing policy changes when receiving a new `LbConfig`, but this is done merely by checking the `currentPolicyName` of the `GracefulSwitchLoadBalancer` component. There is no difference than having a mock/fake `GracefulSwitchLoadBalancer` and verify its `switchTo()` method is called (plus you miss the case when currentLb state is READY before the switch).

Also, this sounds more natural for the test of top-level `XdsLoadBalancer`, which does not do load balancing behavior (it's a wrapper and the actual load balancing logic is performed by its composite components). The behavior of `XdsLoadBalancer` we want to verify is tha it changes the balancer that actually performs load balancing logic upon config update (invocation of `GracefulSwitchLoadBalancer#switchTo()` would suffice).

Reply:
We spoke offline and eventually realized that we don't need GracefulSwitchLoadBalancer in this case. There's going to be some rework of the xds implementation and maybe also some de-mockifying some of the tests and we'll come back to this question of whether we need `SwitchLoadBalancer`.

Reply:
Will rework. Closing.